packages/sdk/client/src/services/worker-client-services.ts

```diff
@@ -0,0 +1,172 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import { Event, Trigger, synchronized } from '@dxos/async';
+import {
+  type ClientServices,
+  type ClientServicesProvider,
+  clientServiceBundle,
+  DEFAULT_CLIENT_CHANNEL,
+} from '@dxos/client-protocol';
+import type { WorkerProxyRuntime } from '@dxos/client-services';
+import type { Stream } from '@dxos/codec-protobuf';
+import { Config } from '@dxos/config';
+import type { PublicKey } from '@dxos/keys';
+import { type LogFilter, parseFilter, log } from '@dxos/log';
+import { type LogEntry, LogLevel } from '@dxos/protocols/proto/dxos/client/services';
+import { createProtoRpcPeer, type ProtoRpcPeer, type ServiceBundle } from '@dxos/rpc';
+import { createIFramePort, createWorkerPort } from '@dxos/rpc-tunnel';
+
+import { IFrameManager } from './iframe-manager';
+import { ClientServicesProxy } from './service-proxy';
+import { ShellManager } from './shell-manager';
+
+/**
+ * Creates services provider connected via worker.
+ */
+export const fromWorker = async (
+  config: Config = new Config(),
+  options: Omit<WorkerClientServicesParams, 'config' | 'createWorker'> = {},
+) => {
+  return new WorkerClientServices({
+    config,
+    createWorker: () =>
+      new SharedWorker(new URL('./shared-worker', import.meta.url), {
+        type: 'module',
+        name: 'dxos-client-worker',
+      }),
+    ...options,
+  });
+};
+
+export type WorkerClientServicesParams = {
+  config: Config;
+  createWorker: () => SharedWorker;
+  shell?: string | null;
+  logFilter?: string;
+};
+
+/**
+ * Proxy to host client service in worker.
+ */
+export class WorkerClientServices implements ClientServicesProvider {
+  readonly joinedSpace = new Event<PublicKey>();
```

Comment:
This seems hacky. What is it used for?

Reply:
nothing anymore, it's removed in a followup 👍